apps/hash-api/src/graph/knowledge/primitive/entity.ts

```diff
@@ -134,9 +139,46 @@ export const createEntity: ImpureGraphFunction<
 export const getEntities: ImpureGraphFunction<
   {
     query: EntityStructuralQuery;
+    temporalClient?: TemporalClient;
   },
   Promise<Subgraph<EntityRootType>>
-> = async ({ graphApi }, { actorId }, { query }) => {
+> = async ({ graphApi }, { actorId }, { query, temporalClient }) => {
+  /**
+   * Convert any strings provided under a top-level 'cosineDistance' filter into embeddings
+   * This doesn't deal with 'cosineDistance' inside a nested filter (e.g. 'any'), so for now
+   * this is only good for single-string inputs.
+   */
+  for (const [filterName, expression] of Object.entries(query.filter)) {
+    if (filterName === "cosineDistance") {
+      if (
+        Array.isArray(expression) &&
+        expression[1] &&
+        "parameter" in expression[1] &&
+        typeof expression[1].parameter === "string"
+      ) {
+        if (!temporalClient) {
+          throw new Error(
+            "Cannot query cosine distance without temporal client",
+          );
+        }
+
+        const stringInputValue = expression[1].parameter;
+        const { embeddings } = await temporalClient.workflow.execute<
+          (params: CreateEmbeddingsParams) => Promise<CreateEmbeddingsReturn>
+        >("createEmbeddings", {
+          taskQueue: "ai",
+          args: [
+            {
+              input: [stringInputValue],
+            },
+          ],
+          workflowId: genId(),
+        });
```

Comment:
Just for my understanding, what's the thinking behind doing this in a temporal worker? My thinking is it makes sense to call an embeddings API directly when generating embeddings as part of a `getEntities` query, so that getting the response isn't blocked on waiting for your workflow to reach the front of the temporal queue.

Reply:
I agree that this may become a problem, and have created H-2017 to follow up. We should fix it before actually using this (e.g. via a search bar)